### ISSUE TITLE
Automatic optical kerning

### ISSUE THREAD

**Issue by adamkrellenstein** [typst/typst#8514]
It would be nice to kern adjacent glyphs from their outlines as a fallback when the font's `kern` table is sparse or absent. Today only the font's own kerning is used; #2692 covers manually-defined pairs and #5826 a case of font kerning going unused, but neither derives kerning from glyph shapes.

Could maybe be a mode on `text`, e.g. `kerning: "metrics" | "optical" | auto`, along the lines of InDesign's metrics/optical switch.

Many otherwise good fonts ship little or no kerning data. Optical kerning would give more even spacing without hand-listing pairs per font.

**Comment by jqwv** [comment]
> It would be nice to kern adjacent glyphs from their outlines 

Do you have an algorithm that can do this? As far as I know there is no public algorithm. But maybe I'm wrong.
The last time I implemented this, I moved both glyphs so near together that they have exactly `MIN_KERN_DISTANCE` euclidean distance at their nearest point. This looks quite good in 80 % of all cases. But some look unoptimized. I have not found any typography article that clearly says how kerning should work and what a good kerning is.

**Comment by adamkrellenstein** [comment]
I think there's a lot out there. The classic is URW's **hz-program** (Karow & Zapf), whose patent has expired; **[[kernagic](https://github.com/hodefoting/kernagic)](https://github.com/hodefoting/kernagic)** is open source and outline-based; and **[[kernall](https://github.com/n8willis/kernall)](https://github.com/n8willis/kernall)** collects the rest.

On `MIN_KERN_DISTANCE`—that's a known approach. The challenge is stem rhythm; ~80% is about the ceiling for pure geometry. There's also a data-driven route: training on existing fonts' spacing ("Learning to Kern", arXiv 2402.14313).

Most of these tools are font editors that emit a kern table, but that's just *when* the analysis runs... nothing stops Typst from doing the same outline analysis itself, computing an optical kern table the first time it sees a font and caching it. That's probably the approach I'd suggest. Applied via `kerning: "optical"` and off by default so it doesn't wreck already-well-fitted fonts.

**Comment by jqwv** [comment]
Thanks for the links and the context. I think the kernagic tool makes it quite clear that it depends on several "strategies" where even one strategy was thrown away in the latest version because the kerning result was too bad. 
The "kindersley method" produces "acceptable results". The "LetterModel" method has limitations because of its assumptions.
This was exactly my assumption: no publicly known good kerning algorithm exists. The method I summarized works rather good 80 % of the time and is simple to understand.
For Typst to implement one method, I think it should be a method that is simple to understand and has a short code. If the code is more than 500 lines, it becomes an own "tool" like the kernagic.
With stem rhythm, this will easily enter the range of 2000 to 3000 lines of code, I guess.

But I agree that such a optical kerning method is really helpful for Typst documents when a font does not have kerning or broken kerning.

**Comment by adamkrellenstein** [comment]
Yes, that totally makes sense.

**Comment by Hyperrick** [comment]
I have been working on an external evaluation for this direction, since the main open question here seems to be whether there is a concrete algorithmic path that would still fit Typst's constraints.

It is not a Typst PR and not a finished API proposal, but a reproducible benchmark suite plus a discussion-ready algorithm candidate called `guarded-profile-hybrid`.

The candidate is:
- deterministic and outline-based,
- applied after shaping, so ligature glyphs are handled as shaped clusters,
- metric-prior, so existing font kerning is preserved where it looks reliable,
- guarded by font-local profile/collision/run-context checks,
- evaluated against Typst metric rendering and scripted InDesign Optical baselines.

Project page:
https://hyperrick.github.io/typst-optical-kerning-bench/

Technical evaluation:
https://github.com/Hyperrick/typst-optical-kerning-bench/blob/main/docs/typst-optical-kerning-evaluation.md

Algorithm notes:
https://github.com/Hyperrick/typst-optical-kerning-bench/blob/main/docs/algorithms.md

I would be interested whether this kind of evidence is useful for the Typst side of the discussion, and which parts would need to be validated next before optical kerning could become a serious compiler topic.

**Comment by jqwv** [comment]
My first impression is that your code seems to work rather good, but for me it is too complex to look through. I think the simple method (nearest euclidean distance between two contours) I proposed above is not a candidate in your approach, maybe it can be a good backup method because it is a clear simple understandable and predictable method while the hybrid combined approach is not predictable for a user, it just means "it does autokern somehow and the result will be often similiar to InDesign". Though, I think your method will probably be better in many cases.

One problem I see is with the proposed API:

```typst
#set text(kerning: "metric")
#set text(kerning: "optical")
#set text(kerning: "none")
``` 

Especially when it comes to "metric". For fonts, (at least) three metric kernings exist:

1) the metrics from the advance widths table
2) the `kern` table
3) the `gpos` table


2 and 3 are optional and can appear together. It is unclear which table is used if kerning is enabled. For example in Windows 11, one font preview uses `kern` and in a different setting it uses `gpos`.
I think it would be good if it's possible to specifiy if `kern` or `gpos` is prefered if both exist.

**Comment by Hyperrick** [comment]
thanks, both points are fair.

the nearest-euclidean-distance method is actually implemented in the workbench as `nearest-contour-distance`. i used it as one of the initial candidates and as a control because it is simple, predictable, and catches real collision cases.

the reason it did not become the leading candidate is similar to what you described: it works well for many pairs, but struggles with things such as apertures, rounded forms, serifs, punctuation, and spacing rhythm across a complete word. i agree that it should remain visible as a simple baseline and possibly as a conservative fallback.

i also agree that the guarded hybrid is currently too complex to propose as-is for typst. the repository contains multiple algorithms, diagnostics, rendering code, and experimental rules. i am not suggesting that this whole implementation should be moved into typst. the next technical step would be to extract the smallest runtime kernel and compare its behavior, complexity, and performance against the nearest-distance and fallback-only methods.

your point about the api is also correct. `metric` is too vague when advance widths, the legacy `kern` table, and gpos positioning can all contribute to the final result. i have removed the earlier `metric | optical | none` sketch from the public page.

the current candidate treats the final positions produced by rustybuzz as the metric prior. in a typst implementation, i think the optical pass should similarly start from the existing shaped glyph positions, rather than independently choosing between `kern` and gpos. whether typst should expose table preference separately is an upstream api decision.

i have written down these open questions and a possible staged path here:
https://github.com/Hyperrick/typst-optical-kerning-bench/blob/main/docs/path-to-typst.md

thanks for looking through it. the complexity question is probably the most important thing to solve before this can become a realistic typst prototype.

**Comment by Francesco146** [comment]
hi everyone, i'm jumping in to share my perspective on this. i want to heavily move to Typst for papers, presentations, and posters, and an automated "optical kerning" fallback would be a massive quality-of-life improvement for my use case. while standard metric kerning works just fine for the 10pt body text of an academic paper, it frequently breaks down in display contexts. my main workflows are these:

1. academic posters (A0 / Large Format): when using standard open-source academic fonts (like Computer Modern, Libertinus, or STIX), the sparse metric kerning tables become obvious at 80pt+ sizes. right now, i have to manually insert horizontal spacing adjustments into poster titles just to make them look balanced
2. presentations (my use case is touying but any really): similar to posters, presentation title slides and large section headers often suffer from gaps, particularly when combining uppercase letters with lowercase typography. this is mostly fine since it's only one slide, but still, it could be better.
3. acronyms and project names:
font foundries rarely include comprehensive metric kerning for these edge cases. an outline-based optical fallback would automatically fix these gaps without requiring us to define custom manual kerning pairs for every new project name we invent.


i strongly support the direction @Hyperrick, even if the initial implementation is just simplified, it's the a step in the right direction, that later can progress in a polished way

**Comment by Hyperrick** [comment]
Following up on the extraction step I mentioned above: I have now completed a smaller compiler-facing runtime kernel and integrated it into a separate Typst prototype.

Prototype commit:
https://github.com/Hyperrick/typst-upstream/tree/908e89562d72a6b27fc903a67584dbc0fffca3e4

Results and architecture:
https://github.com/Hyperrick/typst-optical-kerning-bench/blob/main/docs/runtime-prototype-results.md

The experiment preserves the current `kerning: true` and `kerning: false` behavior and adds `kerning: "optical"` only as a prototype API. The optical pass:

- runs after shaping and starts from Typst's shaped glyph positions;
- treats existing metric positioning as the protected prior;
- handles ligatures as the replacement glyphs produced by shaping;
- uses deterministic outline profiles and memoized font/pair geometry;
- currently limits itself to same-font, left-to-right Latin text;
- adds no new Typst runtime dependency.

The compiler prototype matches the workbench's expected width corrections on all 30 no-ligature and 31 ligature-capable cases, with a maximum absolute difference below `0.000049em`. Its metric-mode PDF is byte-identical to the same-commit baseline. On my local 120-page/240-heading workload, enabling optical kerning only for headings added about `48.4ms` to the median compile time; normal metric-only timing showed no material measured regression.

The compact candidate's current mean rendered difference from the scripted InDesign Optical reference is `0.0200em` on the no-ligature suite and `0.0132em` on the ligature-capable suite. InDesign remains a comparison reference, not ground truth.

Before turning this into an upstream PR, I would appreciate guidance on three boundaries:

1. Is a conservative post-shaping adjustment worth pursuing, or would a missing-kerning-only fallback be preferable?
2. Is left-to-right Latin display text an acceptable first supported slice?
3. Should API shape and the acceptable complexity budget be settled in this issue first, or would a small draft prototype PR be useful for concrete review?

This is still an experiment rather than a merge proposal, and I would keep the benchmark and InDesign tooling outside the compiler patch.

**Comment by jqwv** [comment]
> InDesign remains a comparison reference, not ground truth.

For a ground truth you could maybe use a subset of the [google fonts](https://github.com/google/fonts/tree/main/ofl) that have kerning. Then choose glyph pairs for which kerning data exists and calculate the optical kerning (ttf-parser can extract the kernings, I think). Then compare the results. A txt file with `fontfileID|glyphnameA|glyphnameB|availableKernValue|calculatedKernValue` would be good. We can assume that a specific amount of these fonts has [bad kerning data](https://github.com/google/fonts/issues/6564). But the majority of these fonts has acceptable kerning data. Maybe the 100 most extreme differencies could then be analyzed by hand.

This would give a strong quality impression and is rather simple to derive. It would still require a few hours of work probably. So I can understand if this is too much work. I just wanted to add this possibility of evaluation.

**Comment by Hyperrick** [comment]
Thanks, this was a very useful suggestion. I implemented the audit against the same compact candidate and then used the result to tighten the prototype.

The audit covers 15 pinned Google Fonts, 82 Latin/figure/punctuation characters, 100,860 ordered candidate pairs, and 15,271 pairs with effective shaped kerning. The TSV contains the font id, glyph names and ids, available kern value, calculated value, difference, and sign-change flag.

The initial candidate had 659 sign changes, a 0.0800em 95th percentile, and a 0.2720em maximum difference. The largest differences were indeed concentrated around deliberate positive positioning, especially Pacifico punctuation.

I therefore added a generic bounded metric prior: nonzero kerning cannot change sign, sub-dead-zone values are preserved exactly, and larger values allow at most 0.012–0.030em movement around the shaped value. This is enforced after pair and run adjustments and uses no font or glyph exceptions.

On the unchanged audit, sign changes are now 0, and both the 95th percentile and maximum are 0.0300em. I also reran the visual and compiler suites and documented the small loss of InDesign Optical agreement rather than hiding it.

Report: https://github.com/Hyperrick/typst-optical-kerning-bench/blob/main/docs/metric-preservation-results.md  
Full audit: https://github.com/Hyperrick/typst-optical-kerning-bench/blob/main/docs/metric-agreement-audit.md  
Updated Typst prototype: https://github.com/Hyperrick/typst-upstream/tree/071a3e87b8ccc8d85049d85f31ceb186c949b6a9

I would be interested whether this bounded policy is closer to the preservation behavior you had in mind, or whether you would still prefer a strict missing-kerning-only first scope.

**Comment by laurmaedje** [comment]
@Hyperrick All the work you are posting here seems to be fully AI generated. We do not accept contributions by an AI model anyways, but throwing this amount of AI generated text (the whole repository) at maintainers asking for feedback is a bit disrespectful.

**Comment by Francesco146** [comment]
i did have the same impression, mostly from the comments in this issue and reports. but looking strictly at the methodology, the work behind it seems clean enough and thought. even if this specific codebase is unmergeable under typst's guidelines, i hope the core concept doesn't get entirely discarded - it's an open issue after all. the proposed geometric approach is worth reviewing IMO. perhaps this repository can simply be treated as a PoC? or a design document? if a _human_ contributor is ever interested in this issue, this gives them a sort-of-proven algorithm to study and implement this cleanly

**Comment by Hyperrick** [comment]
Hello @laurmaedje, yes, I used AI to implement the test suite. With a pair of fresh eyes you're right, the artifacts are very extensive and a lot to process. Only to clarify: this was not a "YOLO"-mode result of me prompting an LLM to "fix the issue". I designed the test cases, built and checked the indesign comparisons, evaluated the rendered results, looked for similar real-world use cases. imo this tackles a real world problem, I also tried to find more users on reddit that have similar issues and on top of that, discussed with a typographer about kerning, algorhitms and possible solutions. 

As a side note, a similar AI-assisted workflow recently helped me identify and validate the production problem behind LaurenzV/krilla#410, which led to the fix in LaurenzV/hayro#1282. still I do understand that this is against the contribtion guidelines.I mention this only to explain why I thought this workflow could produce useful technical research.

I don't have enough Rust and compiler experience to independently rewrite and maintain the implementation without AI assistance. For that reason, I think it is better to stop here than to pretend otherwise.

**Comment by jqwv** [comment]
From the test results we can see that my very first comment here was exactly on point. Most results are probably good, but also some unexpected results appear.
I still think that autokerning is an _unsolved problem_. The only way of implementing this with generic good results is using an AI model (CNN).
Typst could use an ONNX model loader, it would load rather fast, has very few lines of extra code, and give better results than InDesign. But it will blow up the typst binary size, probably by 20 % or more. Personally, I would add autokerning as an external tool, if at all. The only predictable solution is, as said, nearest euclidean distance. But in my opinion this is also not extremely trivial to implement, would require 500 loc too, at least.

From my experience it is not possible to solve this without an AI model. Due to the Google Fonts data source, it would be rather simple to derive a perfect model, but I doubt that including such a big dependency (model loader for inference) is worth the functionality.

**Comment by Francesco146** [comment]
@jqwv i agree with you on the ONNX/CNN front bundling a neural network and inference engine is a bad move. but as an end-user, i think we might be letting "perfect" be the enemy of "good enough" here. adobe indesign has featured optical kerning for over two decades using purely deterministic geometry, well before modern ml was a thing. for posters and slides, we don't need typst to act as a professional type foundry tool that generates flawlessly balanced fonts from scratch. we just need a mechanism to stop acronyms like "V-SLAM" or "AWGN" from having massive, awkward gaps at 80pt sizes when a font's metric tables are sparse. if a full geometric hybrid algorithm is too complex or unpredictable for the core compiler, there seem to be two paths:

- an in-compiler conservative heuristic: the simpler nearest-euclidean distance method you mentioned, but restrict it with hard safety limits (e.g., never allow glyphs to collide, and only intervene if the gap exceeds a certain threshold). even if it only fixes the worst 80% of gaps, that is still a good first step.
- external tools and custom pairs: if the basic outline analysis is too heavy for typst, then issue #2692 becomes the blocker. if typst natively allowed us to pass a dictionary like `#set text(kerning-pairs: (("V", "-"): -0.05em))`, then tools like the one in this repository can just exist externally. i could run my poster titles through an external CLI or web app, and it would spit out the exact typst code snippet for me to paste into my documents.

either way it's fine for me and my workflow
